internal/infrastructure/ceph/rbd.go

```diff
@@ -267,6 +357,304 @@ func applyDiffDataRecords(
 	}
 }
 
+func applyDiffDataRecordsWithChecksumVerification(
+	rawFile *os.File,
+	diffFileReader *bufio.Reader,
+	expansionUnitSize uint64,
+	rawChecksumChunkSize uint64,
+	rawChecksumFile *os.File,
+) error {
+
+	stat, err := rawFile.Stat()
+	if err != nil {
+		return fmt.Errorf("failed to stat destination file: %w", err)
+	}
+	fileSize := stat.Size()
+	prevOffset := uint64(0)
+	prevLength := uint64(0)
+
+	// Pre-compute zero checksum for efficiency
+	csumZero := calcZeroChecksum(rawChecksumChunkSize)
+
+	for {
+		tag, err := diffFileReader.ReadByte()
+		if err != nil {
+			if errors.Is(err, csumreader.ErrChecksumMismatch) {
+				return csumreader.ErrChecksumMismatch
+			}
```

Comment:
This code discards the original error, which potentially has additional information. It seems sufficient to return its wrapped error (as in L.385).